### DISCUSSION TITLE
How to programatically deploy functions in a go server inside a kubernetes pod?

### DISCUSSION THREAD

**Discussion by roskee** [knative/func#2139] [General]
There is a [serving](https://github.com/knative/serving/tree/main/pkg/client/clientset/versioned/typed/serving/v1) go package that can be used to deploy knative services programatically from a go code which itself resides in the kubernetes cluster. I am wondering if this is possible with knative functions as well? I tried to use the modues inside the [func/pkg](knative.dev/func/pkg) folder but It seems that these were designed to be used with the func client and they are not supposed to be run inside a cluster but on the machine hosting the cluster. I tried them but ended up with an error saying `cannot create docker client: docker/podman API not available`. am I totally wrong or is this not available yet? Thanks. (I am using a `Kind` cluster on my local machine)

**Comment by lkingland** [comment]
You are right that the client library is currently not able to be run from within a cluster without special configuration (docker within docker).  However, we are working on a builder which does not require a daemon connection in order to perform the build step.  This new builder is available for Go functions, and Python is coming shortly.  Please post if you would like more information about how to use this experimental feature.  If you're targeting other languages, it will be a longer wait.

However your method is correct. you can import the `Client` from `pkg/functions` and instantiate it with some concrete implementations of your choice for `builder` and `deployer`.  Once the aforementioned daemon-less builder is available we will be updating our documentation, both for the cli and for the Go library.
